### PR TITLE
WIP: Add oauth proxy to ODH dashboard

### PR DESCRIPTION
**Pull request by accorvin**
This adds a basic OAuth Proxy sidecar container to the ODH Dashboard
pod. By default this will allow any authenticated user to access the
service. I expect downstream deployments to limit access by way of the
--openshift-sar argument. I additionally expect downstream deployments
to customize the value of the --cookie-secret argument.

### CODE REVIEW

File: odh-dashboard/base/deployment.yaml [hunk 1]
```diff
@@ -14,6 +14,31 @@ spec:
...
       containers:
+      - name: oauth-proxy
+        args:
+        - --https-address=:8443
+        - --provider=openshift
+        - --openshift-service-account=odh-dashboard
+        - --upstream=http://localhost:8080
+        - --tls-cert=/etc/tls/private/tls.crt
+        - --tls-key=/etc/tls/private/tls.key
+        - --cookie-secret=SECRET
```
Comment (vpavlin) [inline odh-dashboard/base/deployment.yaml:25]: Do we need to parameterize te SECRET?

Comment (accorvin) [reply odh-dashboard/base/deployment.yaml:25]: likely, yes. I could use a pointer to an example for how to do that.

Downstream deployments will use a slightly different method to generate a random value of this (see https://github.com/red-hat-data-services/odh-deployer/blob/main/deploy.sh#L34)

**Comment by vpavlin** [comment]
Should we have access controlled by groups here as well?

I think we could use https://github.com/openshift/oauth-proxy#require-specific-permissions-to-login-via-oauth-with---openshift-sarjson

And define: https://docs.openshift.com/container-platform/4.5/rest_api/authorization_apis/subjectaccessreview-authorization-openshift-io-v1.html

I am not sure how to define the list of groups though as they are in JH ConfigMap and here we'd probably have to do it via parameters

File: odh-dashboard/base/deployment.yaml (hunk 1 above)
Comment (vpavlin) [reply odh-dashboard/base/deployment.yaml:25]: Hmm...we do not have a way to generate a random value at the moment.

Here is an example of a parameter:
- https://github.com/opendatahub-io/odh-manifests/blob/master/jupyterhub/jupyterhub/base/jupyterhub-singleuser-profiles-configmap.yaml#L11
- https://github.com/opendatahub-io/odh-manifests/blob/master/jupyterhub/jupyterhub/base/params.env#L2
- For non-standard fields you need to add kind and path here: https://github.com/opendatahub-io/odh-manifests/blob/master/jupyterhub/jupyterhub/base/params.yaml
- https://github.com/opendatahub-io/odh-manifests/blob/master/jupyterhub/jupyterhub/base/kustomization.yaml#L43-L49 (the CM `parameters` is generated based on `params.env`)

**Comment by crobby** [comment]
I think the tests will likely fail unless they are updated a bit:  https://github.com/opendatahub-io/odh-manifests/blob/master/tests/basictests/dashboard.sh#L29

File: odh-dashboard/overlays/authentication/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+- op: add
+  path: /spec/template/spec/containers/-
+  value:
+    - name: oauth-proxy
```
[the remark this answers is not in the document]
Comment (accorvin) [reply odh-dashboard/overlays/authentication/deployment.yaml]: Yep, I'm still working on testing this. I'll let you know once I have it working again and this is ready for reviews once more.

**Comment by accorvin** [comment]
Closing until I get the tests working, will reopen.

**Comment by accorvin** [comment]
I've reopened this branch for PR in #350 

All comments here should be resolved there (and I'm hoping the tests will pass)

**Pull request closed without merging**